Navigation: Fix click-button size, submenu directions, scrollbars.

jasmussen opened the pull request:
## Description

Fixes https://github.com/WordPress/twentytwentytwo/issues/206. As reported there, there are a few issues with dropdown submenus and their nested versions as well:

<strike>- Even when the submenu is closed, its contents take up space and can in some circumstances create a horizontal scrollbar.</strike>
- Submenus should open leftwards when right aligned. This worked, but not at small breakpoints.
<strike>- Not reported in that ticket, but surfaced by it — menus set to open on click had smaller buttons for the submenus than for hover.</strike>
- The contents of burger menus should always be top aligned.

Some of the bugs originally fixed by this PR have been fixed separately. But the two remaining issues are still good to land.

## How has this been tested?

<strike>
- Create navigation menus that features submenus, and sub sub menus (nested menus).
- Test with and without click enabled.
- Test a justified-right submenu on small (<782) and bigger breakpoints. They should both open leftwards, but <782 menus should open leftwards and _downwards_ (as shown in the GIF above).
- Bonus: use the web inspector to verify that submenus are zero width even when not shown.

Note that we use `visibility: hidden:` to hide submenus so we can nicely fade them in and out. `display: none;` is an alternative that won't let us animate it as nicely.</strike>

Test instructions and GIFs are shown in https://github.com/WordPress/gutenberg/pull/36215#issuecomment-966174501.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

jasmussen committed:
Navigation: Fix click-button size, submenu directions, scrollbars.

jasmussen commented on packages/block-library/src/navigation/style.scss:
```diff
@@ -130,6 +130,7 @@
 		// Don't take up space when the menu is collapsed.
 		width: 0;
 		height: 0;
+		overflow: hidden; // Overflow is necessary to set, otherwise submenu items will take up space.
```
This overflow rule I believe I had in the past, and then removed it. In this PR it appears to work as intended, with no ill side effects. But I would appreciate thorough testing of submenus, with and without click, and with and without keyboard usage. Just to make sure I'm not missing anything.

vcanales reviewed (commented):
The right alignment fix works but only when explicitly aligned right; it doesn't seem to fix the problem when the navigation menu is used in a template, where it happens to sit at the right of the screen.

![Kapture 2021-11-04 at 13 45 13](https://user-images.githubusercontent.com/1157901/140382582-3998f0c7-6adf-47b8-bf40-875584f13ae3.gif)

It's also noticeable that long submenu items might overflow outside of the viewport; is it possible to wrap them conditionally? ie. when they exceed the viewport.

<img width="641" alt="image" src="https://user-images.githubusercontent.com/1157901/140384584-0cc8418c-2c7d-4fca-bf75-149a56202238.png">

Other than that, I didn't find any other details! 👍

jasmussen commented:
Thanks so much for the thorough review! I'll look at addressing this.

jasmussen commented:
> The right alignment fix works but only when explicitly aligned right; it doesn't seem to fix the problem when the navigation menu is used in a template, where it happens to sit at the right of the screen.

Yes, that's a bit of a limitation of how things work at the moment, but beyond the scope of this PR to fix. I created #36241 to track this further.

jasmussen commented:
> It's also noticeable that long submenu items might overflow outside of the viewport; is it possible to wrap them conditionally? ie. when they exceed the viewport.

It appears I intentionally added nowrap to menu items in https://github.com/WordPress/gutenberg/pull/35077. I'll see if I can remove it again, but I think it was necessary due to some scaling issues.

It also surfaced some other issues, which I'll look into.

jasmussen committed:
Fix wrapping, fix z index issue.

jasmussen commented:
Fixed two issues. First off, the explicit application of a z-index when the mobile responsive menu is enabled caused this:
<img width="546" alt="bug" src="https://user-images.githubusercontent.com/1204802/140479642-d1729548-b24c-48a3-a2b2-b6af10370626.png">

Setting it to "auto" fixed it:
<img width="651" alt="bugfix" src="https://user-images.githubusercontent.com/1204802/140479656-c776d008-d0a3-4535-946c-84f3ed434d3e.png">

I also removed the `nowrap`, because even without it, I can no longer reproduce the issue I saw in https://github.com/WordPress/gutenberg/pull/35077

jasmussen committed:
Fix issue with justification and responsive container.

jasmussen commented:
Alright, fixed a last issue with justificaitons broken when mobile responsiveness was enabled.

<img width="800" alt="Screenshot 2021-11-05 at 09 27 29" src="https://user-images.githubusercontent.com/1204802/140480991-2db994fb-16cb-46f7-99ae-d7e68369cf3a.png">
 
This one should be good to go now.

jasmussen commented:
This one needs a good rebase after the merging #36169.

vcanales commented:
> This one needs a good rebase after the merging #36169.

By itself, it's looking good to me. Rebase-wise, there was only one conflict to fix:

```css
.wp-block-navigation__responsive-container-content {
<<<<<<< HEAD
  display: contents;
=======
  display: flex;
>>>>>>> 329693fea1 (Fix issue with justification and responsive container.)
}
```

`display: flex` was replaced by `display: contents` in #36169. I'll push the rebased branch in order to move forward :)

jasmussen commented:
Thank you for the rebase help. It might need more changes, though, as 36169 removed the justification classes, so menus no longer open in the correct direction at all. I think @tellthemachines had some ideas to explore here as well.

jasmussen commented:
I've extracted some of the small fixes (line height, overflow etc) into #36298, which can likely land easily as it's small. This PR should rebase right over it, but is still blocked on either restoring the justification classes or finding another way forward. But by extracting the other things, this one can focus up a bit.

jasmussen commented:
Alright, I think this one is ready for review again. A lot of what this PR originally fixed has now been made moot as it was extracted and fixed separately, so there are primarily two things left to test, and both in my own testing appear fixed by this PR.

1. With a justified-right navigation menu that has multiple levels of nesting, verify that menus on desktop breakpoints open down and leftwards. On mobile breakpoints, they open directly downwards:
![resizing](https://user-images.githubusercontent.com/1204802/141279595-18d7b196-28c6-4262-bd67-490e201fb33e.gif)

2. With a justified-right (or centered) navigation menu set to always be burgered, verify that menus are always aligned to the top, even if they are left/centered/right justified:
<img width="892" alt="Screenshot 2021-11-11 at 11 07 17" src="https://user-images.githubusercontent.com/1204802/141279799-d7d52723-e83e-44f4-9e25-d129a49d0ad8.png">

jasmussen committed:
Remove duplicate rule.

tellthemachines commented:
@jasmussen I can no longer reproduce either of those issues on trunk. To clarify, were the submenus on small breakpoints not opening to the bottom? I've been comparing trunk to this branch and not quite sure what has changed 😅

jasmussen commented:
Sorry the editing of the original issue actually removed helpful instructions. I still see this happening on trunk:

<img src="https://user-images.githubusercontent.com/1204802/140333085-21a06e6d-c18a-4bbf-a41c-12acb3e8135d.gif">

That's becasue the rules for justifying the menus are wrapped in a media query. So the steps to test:

- Create navigation menus that features submenus, and sub sub menus (nested menus).
- Set them to open on click, set them justified right.
- Open the submenu 2 levels deep
- Resize the window smaller than 782px. 

At around that breakpoint, the menus should in this PR start to open _downwards_, but still be right aligned to their parent item. In trunk, they just open rightwards at that point.

tellthemachines commented:
Oooh got it, thanks. I was looking at the nested submenus and ignoring where they were sitting at the top level 🤦 

I think this is good to go!

jasmussen commented:
Thank you!
